src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs

```diff
@@ -0,0 +1,100 @@
+        /// </summary>
+        /// <param name="loaderContextAccessor">The <see cref="IAssemblyLoadContextAccessor"/>.</param>
+        /// <param name="assemblyProvider">The <see cref="IAssemblyProvider"/> that provides assemblies
+        /// for precompiled view discovery.</param>
+        public DefaultPrecompiledViewsProvider(IAssemblyLoadContextAccessor loaderContextAccessor,
+                                               IAssemblyProvider assemblyProvider)
+        {
+            _loadContext = loaderContextAccessor.GetLoadContext(RazorFileInfoCollectionType.Assembly);
+            _assemblyProvider = assemblyProvider;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<PrecompiledViewInfo> PrecompiledViews
+        {
+            get
+            {
+                var precompiledViews = new List<PrecompiledViewInfo>();
+                var razorFileInfoCollections = GetRazorFileInfoCollections();
+                foreach (var viewCollection in razorFileInfoCollections)
+                {
+                    var viewAssembly = LoadPrecompiledViewsAssembly(viewCollection);
+                    foreach (var fileInfo in viewCollection.FileInfos)
+                    {
+                        var viewType = viewAssembly.GetType(fileInfo.FullTypeName);
+                        precompiledViews.Add(new PrecompiledViewInfo(fileInfo, viewType));
+                    }
+                }
+
+                return precompiledViews;
+            }
+        }
+
+        private IEnumerable<RazorFileInfoCollection> GetRazorFileInfoCollections()
+        {
+            return _assemblyProvider.CandidateAssemblies
+                                    .SelectMany(a => a.ExportedTypes)
+                                    .Where(Match)
+                                    .Select(c => (RazorFileInfoCollection)Activator.CreateInstance(c));
+        }
+
+        // Internal for unit testing.
+        internal static bool Match(Type type)
+        {
+            var typeInfo = type.GetTypeInfo();
+            if (RazorFileInfoCollectionType.IsAssignableFrom(typeInfo))
+            {
+                var hasParameterlessConstructor = type.GetConstructor(Type.EmptyTypes) != null;
+
+                return hasParameterlessConstructor &&
+                       !typeInfo.IsAbstract &&
+                       !typeInfo.ContainsGenericParameters;
+            }
+
+            return false;
+        }
+
+        private Assembly LoadPrecompiledViewsAssembly(RazorFileInfoCollection viewCollection)
+        {
+            var viewCollectionAssembly = viewCollection.GetType().GetTypeInfo().Assembly;
+            using (var assemblyStream = viewCollectionAssembly.GetManifestResourceStream(
+                viewCollection.AssemblyResourceName))
+            {
+                Stream symbolsStream = null;
+                if (!string.IsNullOrEmpty(viewCollection.SymbolsResourceName))
```

Comment:
should we generate pdbs for release builds? Probably not I assume, because it will make the project assembly much larger